### PR TITLE
fix(core): refactor to work with Sails 1.0, OpenAPI 3, JSDoc (work in progress)

### LINKED ISSUE #11: Not generating routes from controller correctly
Hi I get something along the lines of 

No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'

When I sails lift. These are related to the functions in my controller. Is there anyway to have the swagger document generated without having to directly specify the routes again in config.routes.js 

like so

  ```
'get /v1/vdc/provisionNew': {
    controller: 'v1/vdcController',
    action: 'provisionNew',
    swagger: {
      summary: 'ProvisionNew',
      description: 'This is for authentication of any user',
      body: {
        username: { type: 'string', required: true },
        password: { type: 'password', required: true }
      },
      query: [{ $ref: '#/parameters/IDPathParam' }]
    }
  }
```

It would be better if it could autogenerate from the controller without having to redo the routes.

### LINKED ISSUE #39: Re-write to better support sailjs v1
related to the following 
 - #11 
 - #18 
 - #28

### LINKED ISSUE #51: Disable specific routes to generate
Hi Guys,

Can we disable some specific routes to generate in swagger?

Thanks

### PR DESCRIPTION
**Pull request by danielsharvey**
Refactor hook initialisation to run after Sails ready in order to
ensure that `sails.getActions()` returns appropriate content.

Replace removed `sails.controllers` with equivalent data structure
derived from `sails.getActions()` (work in progress).

Utilise model attribute `columnType` to infer more type info (work
in progress).

Addresses https://github.com/theo4u/sails-hook-swagger-generator/issues/11#issuecomment-437683002 and https://github.com/theo4u/sails-hook-swagger-generator/issues/39

closes #11 
closes #39 
closes #51

### CODE REVIEW

**Comment by danielsharvey** [comment]
This needs further checking to ensure it is complete and covers all aspects of mapping actions and model attributes, but I've created a PR in case of interest and for comment on direction.

**Comment by theo4u** [comment]
@danielsharvey thanks.

> Replace removed sails.controllers with equivalent data structure
derived from sails.getActions() (work in progress). 

See my thoughts about it here: https://github.com/theo4u/sails-hook-swagger-generator/issues/18#issuecomment-370853338

File: lib/swaggerDoc.js [hunk 1]
```diff
@@ -14,7 +14,25 @@ module.exports = function (sails, context) {
     specifications.parameters = generators.parameters(sails.config, context);
 
 
-    var generatedRoutes = generators.routes(sails.controllers, sails.config, specifications.tags);
+    let actions = sails.getActions();
```
Comment (theo4u) [inline lib/swaggerDoc.js]: +1

**Comment by theo4u** [comment]
With this, there's no support for sails v0.12.x and maybe we want to do some fixes or feats that cuts across both version, I don't think it will be possible again with

```diff
-    "sails": "~0.11.0",
+    "sails": "^1.1.0"
```


Wdyt?

**Comment by danielsharvey** [comment]
Apologies for long delay responding. I suspect maintaining support for both 0.12.x and 1.x will be difficult in the same release. Maintaining 2.x for Sails v0.12.x and releasing 3.x for 1.x support might be pragmatic.

> With this, there's no support for sails v0.12.x and maybe we want to do some fixes or feats that cuts across both version, I don't think it will be possible again with
> 
> ```diff
> -    "sails": "~0.11.0",
> +    "sails": "^1.1.0"
> ```
> 
> Wdyt?

Will push updated support shortly.

**Comment by danielsharvey** [comment]
@theo4u, by way of update: This is progressing although not as quickly as I would like.

Question: Do you have an opinion on moving from Swagger to to OpenAPI 3?

**Comment by theo4u** [comment]
> Question: Do you have an opinion on moving from Swagger to OpenAPI 3? 

No, I've not given this a thought

**Comment by danielsharvey** [comment]
@theo4u Quick update as this has taken longer than anticipated. I found issues with using `sails.getActions()` and have moved to using listening to `router:bind` [events](https://github.com/balderdashy/sails/blob/master/lib/EVENTS.md#routerbind). I've also worked to upgrade output to OpenAPI 3.

**Comment by theo4u** [comment]
@danielsharvey cool 👍 , can you push so I try it out

**Comment by danielsharvey** [comment]
@theo4u, this turned out to be a more significant refactor that I expected 
once I got started now that I have included Sails 1.0, OpenAPI 3 and JSDoc.

Updates to the tests are a work-in-progress - I have not updated the all tests, only
the bootstrap test so far.

To run bootstrap only, use:

```bash
node ./node_modules/mocha/bin/mocha test/bootstrap.test.js
```

I have tested quite extensively on a project I'm using this for, but the test output
produced in `swagger/swagger.json` by the bootstrap test is OK, and has been
tested against:
- http://editor.swagger.io/
- https://redocly.github.io/redoc/

**Comment by eduardotamaki** [comment]
Hi @danielsharvey, just tried your code and despite some warnings, it worked to generate the open api from blueprints of my sails 1.0 application. 

Keep up with the good work!

**Comment by danielsharvey** [comment]
Thanks for the feedback @eduardotamaki. Feel free to let me know any specific warnings - I've dealt with a reasonable number so far but happy to address (as noted, I know that work is needed to finished updating the tests).

**Comment by theo4u** [comment]
> @theo4u, this turned out to be a more significant refactor that I expected
> once I got started now that I have included Sails 1.0, OpenAPI 3 and JSDoc.
> 
> Updates to the tests are a work-in-progress - I have not updated the all tests, only
> the bootstrap test so far.
> 
> To run bootstrap only, use:
> 
> ```shell
> node ./node_modules/mocha/bin/mocha test/bootstrap.test.js
> ```
> 
> I have tested quite extensively on a project I'm using this for, but the test output
> produced in `swagger/swagger.json` by the bootstrap test is OK, and has been
> tested against:
> 
> * [http://editor.swagger.io](http://editor.swagger.io/)
> * [redocly.github.io/redoc](https://redocly.github.io/redoc/)

Will try it out today and wait for the tests so we know how to release this, in case of any breaking changes

**Comment by EvanTimmermann** [comment]
Hey @theo4u and/or @danielsharvey,
Has this PR been looked at recently?  I just stumbled across it and I'm interested in seeing this implemented.

**Comment by danielsharvey** [comment]
> Hey @theo4u and/or @danielsharvey,
> Has this PR been looked at recently? I just stumbled across it and I'm interested in seeing this implemented.

@EvanTimmermann I regularly use this internally but I've not worked on the tests unfortunately. I will have some space in several weeks to look at it.

**Comment by enricodeleo** [comment]
I was having problems with the original branch with sails 1.2.3. None of the blueprint rest endpoints was documented. 

With this fork, everything works like a charm so THANKS! 

I can't wait to see this merged!

**Comment by theo4u** [comment]
Will try looking at this coming weekend so we can merge it in

**Comment by danielsharvey** [comment]
Thanks for the feedback @enricodeleo. As I mentioned previously, I've not had time to work on the automated tests. But I will make sure I contribute quickly if other issues found (and I can help!) - I've been using myself for quite some time.

**Pull request closed without merging**